In [1]:
import pandas as pd

In [16]:
def get_attendance_grade(letter):
    if letter == 'P':
        return 1
    elif letter == 'A':
        return 0
    elif letter == 'T':
        return 0.9
    elif letter == 'H':
        return 0.5

get_attendance_grade('H')

0.5

In [23]:
grades = {
    'P': 1,
    'A': 0,
    'H': 0.5,
    'T': 0.9,
}

df = pd.read_csv('untidy-data/attendance.csv')
df = df.rename(columns={'Unnamed: 0': 'student'})
df = df.melt(id_vars='student', var_name='date')
df['grade'] = df.value.apply(get_attendance_grade)
df.groupby('student').grade.mean()

student
Billy    0.5250
Jane     0.6875
John     0.9125
Sally    0.7625
Name: grade, dtype: float64

In [35]:
df = pd.read_csv('untidy-data/coffee_levels.csv')
df = df.pivot_table(values='coffee_amount', index='hour', columns='coffee_carafe')
df.round(3)

coffee_carafe      x      y      z
hour                              
8              0.816  0.189  0.999
9              0.451  0.522  0.916
10             0.843  0.023  0.145
11             0.336  0.236  0.311
12             0.898  0.017  0.772
13             0.311  0.997  0.399
14             0.507  0.058  0.864
15             0.215  0.145  0.436
16             0.184  0.545  0.281
17             0.392  0.594  0.437

In [51]:
df = pd.read_csv('untidy-data/cake_recipes.csv')
recipes_and_positions = df['recipe:position'].str.split(':', expand=True)
recipes_and_positions.columns = ['recipe', 'position']
df = pd.concat([df, recipes_and_positions], axis=1).drop(columns='recipe:position')
df = df.melt(id_vars=['recipe', 'position'], var_name='temp', value_name='tastiness')
df

recipe position temp  tastiness
0       a   bottom  225  61.738655
1       a      top  225  51.709751
2       b   bottom  225  57.095320
3       b      top  225  82.455004
4       c   bottom  225  96.470207
5       c      top  225  71.306308
6       d   bottom  225  52.799753
7       d      top  225  96.873178
8       a   bottom  250  53.912627
9       a      top  250  52.009735
10      b   bottom  250  61.904369
11      b      top  250  95.224151
12      c   bottom  250  52.001358
13      c      top  250  82.795477
14      d   bottom  250  58.670419
15      d      top  250  76.101363
16      a   bottom  275  74.414730
17      a      top  275  68.576858
18      b   bottom  275  61.196980
19      b      top  275  98.594881
20      c   bottom  275  92.893227
21      c      top  275  92.098049
22      d   bottom  275  51.747686
23      d      top  275  59.571620
24      a   bottom  300  98.786784
25      a      top  300  50.228470
26      b   bottom  300  99.248541
27      b      top  300  58.169349
28      c   bottom  300  65.473084
29      c      top  300  53.960273
30      d   bottom  300  56.183110
31      d      top  300  50.971626

Which recipe, on average, is the best? recipe b

In [52]:
df.groupby('recipe').tastiness.mean()

recipe
a    63.922201
b    76.736074
c    75.874748
d    62.864844
Name: tastiness, dtype: float64

Which oven temperature, on average, produces the best results? 275

In [53]:
df.groupby('temp').tastiness.mean()

temp
225    71.306022
250    66.577437
275    74.886754
300    66.627655
Name: tastiness, dtype: float64

Which combination of recipe, rack position, and temperature gives the best result? recipe b, bottom rack, 300 degree

In [54]:
df.sort_values(by='tastiness').tail()

recipe position temp  tastiness
4       c   bottom  225  96.470207
7       d      top  225  96.873178
19      b      top  275  98.594881
24      a   bottom  300  98.786784
26      b   bottom  300  99.248541

Other methods to look at:

- `.pivot`
- `.stack`
- `.unstack`